### PR TITLE
App rewriting

### PR DESCRIPTION
**Pull request by PatrykSzwer**
Reference to #80, this will also implement #79 

@Joltrix you can check if current version is working with Timeline app. I think everything should be okay, but let's be 100% sure. 

It's still not merge'able yet, I'll be adding more changes to this PR. For description of most of the initial changes, please read commit message of first commit.

### CODE REVIEW

**Comment by PatrykSzwer** [comment]
There is a big, complex problem with bindings in calendar. I'll write steps to reproduce it with this branch:

1. Select any event / create new event and click edit.
2. go to participants and start writting "adm"
3. You will find that there will be no proposition of person in view but You'll find that Admin person was found in debug aid, in view-model and also in patches.

Other easy way to reproduce It will be:

1. Go to EventPage.html and uncomment back button (89 line) and comment <a> link.
2. Add label with `{{model.RedirectUrl$}}` for test purposes.
3. Click on Back button.
Again, label on page will not be changed, but value in View-model/patches will be updated.

**Comment by alshakero** [comment]
It's giving me this:
image


I launched SignIn, but I don't know how to add test user and pass to the DB. Do you know how to do that?

**Comment by PatrykSzwer** [comment]
Sorry, probably I should give standard details about preparation. You have to have a SignIn app running, generated admin user (just go to ` http://localhost:8080/signin/generateadminuser`) and log in into it. (Login: Admin, pass same)

**Comment by PatrykSzwer** [comment]
@Joltrix I'm thinking about changing strategy, and ending this PR with current changes and creating next one explicitly for view models, since those needs really complex changes. So I think You can start checking my changes and giving suggestions about them.

**Comment by alshakero** [comment]
Found the reason why. All details here https://github.com/Starcounter/starcounter-include/issues/52

**Comment by PatrykSzwer** [comment]
@alshakero I've used recommended solution with `imported-template`. However there is new issue that showed up bc of this solution. For some reason, when I'm entering an event view from calendar, I'm getting invalid frame with opCode -1. If You can please take a look. Reproduction steps for this branch with latest commits:

1. Go to http://localhost:8080/Calendar/
2. Sign in if You're not.
3. Select any event / create new event and click edit.

You will get Invalid frame header error.

**Comment by Mihaiii** [comment]
I can access an event without being logged in if I know the URI.
Since the home page requires the user to be logged in, so should the events.

File: src/Calendar/Api/MainHandlers.cs [hunk 1]
```diff
@@ -1,45 +1,44 @@
...
+            Application.Current.Use(new PartialToStandaloneHtmlProvider());
+
+            Handle.GET("/Calendar/app-name", () => new AppName());
+
+            Handle.GET("/Calendar/partial/deny", () => new PermissionDeniedPage());
+
             Handle.GET("/calendar/timeline-item/{?}", (string eventId) =>
             {
-
                 Event thisEvent = DbHelper.FromID(DbHelper.Base64DecodeObjectID(eventId)) as Event;
```
Comment (Joltrix) [inline src/Calendar/Api/MainHandlers.cs]: This can be changed to:

```c#
Event thisEvent = Db.FromId<Event>(eventId);
```

**Comment by PatrykSzwer** [comment]
Thanks for pointing that out Mihaiii, I fiixed it.

File: src/Calendar/wwwroot/Calendar/viewmodels/EventPage.html [hunk 2]
```diff
@@ -87,9 +85,10 @@ <h3 slot="calendar/event-participants-label">Participants</h3>
         </div>
         <button slot="calendar/event-save-button" class="btn btn-primary" disabled$="{{isSaveDisabled(model.IsInvalid, model.IsPristine)}}" onclick="++this.value;" value="{{model.Save$::click}}">Save</button>
         <button slot="calendar/event-undo-button" class="btn btn-default" onclick="++this.value;" value="{{model.Rollback$::click}}">Undo</button>
-        <button slot="calendar/event-delete-button" class="btn btn-warning" onclick="++this.value;" value="{{model.Delete$::click}}">Delete</button>
-        <button slot="calendar/event-close-button" class="btn btn-default" onclick="++this.value;" value="{{model.Close$::click}}">Close</button>
-        <link is="puppet-redirect" history url$="{{model.RedirectUrl}}" />
+        <a></a><button slot="calendar/event-delete-button" class="btn btn-warning" onclick="++this.value;" value="{{model.Delete$::click}}">Delete</button>
```
Comment (Joltrix) [inline src/Calendar/wwwroot/Calendar/viewmodels/EventPage.html]: Is there a reason for the `<a></a>` before the button :O?

**Comment by Joltrix** [comment]
I couldn't find anything that looked odd in the code, good job with the rewriting :)

The only things we should have fixed before we merge this is:
1. The Edit crash, which you have mentioned above.

And the url redirect that we've been talking about. Which happens if one tries to edit an old event (a passed date)
calendarmisfire

**Comment by Mihaiii** [comment]
If I'm not logged in, I get a black page (there used to be `PermissionDeniedPage`).

**Comment by PatrykSzwer** [comment]
As I discovered @Joltrix, and I have completely no idea why, that additional string in URI is added by custom element only when We're using `imported-template`. With starcounter-include there is no additional string added. You can check and confirm if It's happening on your machine also.

**Comment by Mihaiii** [comment]
Please see the gif.
This is a little confusing for a new user. So either the user should be able to edit the interval after he presses "Edit", either the "begin" and "end" values should be simple text to induce the idea that once "edit" is clicked, they are non-changeable (so the interval can be selected only before "edit" is pressed).

edit_time_interval

**Comment by Joltrix** [comment]
Should they be editable though? But I guess it can be difficult to set up, considering you can have a lot of other events "around" that event, which would cause the events to overlap if they were edited from within

**Comment by Mihaiii** [comment]
> Should they be editable though? But I guess it can be difficult to set up, considering you can have a lot of other events "around" that event, which would cause the events to overlap if they were edited from within

Don't know; I went with two options in the comment (editable and non-editable). IMO they should be editable: the same kind of check that is performed before clicking "edit" (to assure is not overlapping with another event) can be performed after it.

**Comment by Mihaiii** [comment]
Oh, there isn't a check for overlapping events event before clicking "edit".
There should be one?

File: src/Calendar/Database/CommitHooks.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ private static void RefreshSignInState()
 
                 var page = session.Store[nameof(MasterPage)] as MasterPage;
 
-                page?.Transaction.Scope(() => page.RefreshCurrentPage());
+                page?.Transaction?.Scope(() => page?.RefreshCurrentPage());
```
Comment (alshakero) [inline src/Calendar/Database/CommitHooks.cs]: The second `?` for `page?` is extraneous I guess. You won't make it there if page is null due to first `page?`

Comment (Joltrix) [reply src/Calendar/Database/CommitHooks.cs]: Yeah, that's true! I'll remove the second question mark :)

**Comment by PatrykSzwer** [comment]
That thing with noneditable hour is a bug. I'm currently working on it.

**Comment by PatrykSzwer** [comment]
Take a look, entire date should be editable right now.

**Comment by Joltrix** [comment]
Yeah, I tested your changes, and the event date & time can be edited!! :)

It did however not solve #82 like I had hoped it would :(

**Comment by PatrykSzwer** [comment]
I've done some more changes and removed more unused files. @Joltrix please take a look if I didn't broke something unintentionally.

**Comment by PatrykSzwer** [comment]
I think after last check we can merge this into dev branch. Next thing will be adding necessary features that should be supported by our calendar, like differences between users, what one person can see/edit and what another, new things to add etc. For now on, please @Joltrix do final review.

**Comment by PatrykSzwer** [comment]
I found out that there came up new issue with transaction. I'll try to fix this.

**Comment by Joltrix** [comment]
I'll wait for you to fix the issue you found, and I'll review and test so that everything works afterwards/tomorrow :)

**Comment by PatrykSzwer** [comment]
I've spoke with @chrhol and turns out that there is a bug in newest SC update. He will fix it and problem with transaction should dissapear.

**Comment by PatrykSzwer** [comment]
Fix for transactions issue is in SC version  2.3.1.7158 (thx again @chrhol for quick response!). In a meantime I've added some additional permission features for events for logged in users (Event is editable only by owner/admin). @Joltrix I think We can finally review it!

**Comment by Joltrix** [comment]
I've tested it again together with the latest develop branches of: SignIn, Timeline, Notes, People

and everything works really smooth ~ :D! There are no longer issues with the creation of events, like it was before where creating new events sometimes wouldn't work due to the calendar not being selectable.
Nor do the issues regarding the update (Save button once something had been updated in the event information) remain.

**Comment by Joltrix** [comment]
Unless @Mihaiii has any objections, I would say that you can go ahead and merge this :) And if anything appears, just add it in as a new issue

**Pull request merged**